### ISSUE TITLE
ECS was unable to assume the role ecsTaskExecutionRole

### ISSUE THREAD

**Issue by arminus** [1Mill/example-terraform-ecs#3]
First of all, thanks for the exhaustive examples, I'm afraid I'm stuck though on examples/terraform-aws-modules/main.tf line 152...

Basically, I've manually created a role ecsTaskExecutionRole in IAM > Roles and assigned AmazonECSTaskExecutionRolePolicy to it.

![2023-10-05_150215](https://github.com/1Mill/example-terraform-ecs/assets/683680/aa2ab4f2-d732-4476-bc0f-c4607ab3ea94)

The trust relation is set to 
```
{
    "Version": "2008-10-17",
    "Statement": [
        {
            "Sid": "",
            "Effect": "Allow",
            "Principal": {
                "Service": "ecs.amazonaws.com"
            },
            "Action": "sts:AssumeRole"
        }
    ]
}
```
and I can retrieve it with `aws iam get-role --role-name ecsTaskExecutionRole`

However, the aws_ecs_task_definition doesn't succeed but reports this as service event for example-ecs-terraform-aws-modules-service:

_service [example-ecs-terraform-aws-modules-service](https://eu-central-1.console.aws.amazon.com/ecs/v2/clusters/example-ecs-terraform-aws-modules/services/example-ecs-terraform-aws-modules-service?region=eu-central-1) failed to launch a task with (error **ECS was unable to assume the role 'arn:aws:iam::<m<id>:role/ecsTaskExecutionRole'** that was provided for this task. Please verify that the role being passed has the proper trust relationship and permissions and that your IAM user has permissions to pass this role.)._

So I dug a little bit deeper into the [IAM docs](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_passrole.html) and tried adding a Customer inline permission policy for my terraform user like so:

```
{
	"Version": "2012-10-17",
	"Statement": [
		{
			"Effect": "Allow",
			"Action": [
				"iam:GetRole",
				"iam:PassRole"
			],
			"Resource": "arn:aws:iam::<m<id>:role/*"
		}
	]
}
```

But that doesn't help either. So some guidance as to how exactly create that ecsTaskExecutionRole and what other preconditions have to be met would be highly appreciated!

**Comment by 1Mill** [comment]
Hey @arminus, 

This IAM Role used to be automatically created by AWS by default, however, they have since stopped doing that hence the problem and `TODO` I added to the code and haven't got around to fixing yet in this repo. Thanks for reminding me. 

Nonetheless, here is the snippet from another workshop I ran to create the IAM Role that AWS doesn't automatically create any more:

```hcl
# * --- Step 14 ---
# * Next, we need to give our cluster permission to run our application.
# *
# * We do this by creating an IAM Role with the AWS managed "AmazonECSTaskExecutionRolePolicy"
# * which will allow our cluster to do ECS related things.
# *
data "aws_iam_policy_document" "this" {
	version = "2012-10-17"

	statement {
		actions = ["sts:AssumeRole"]
		effect = "Allow"

		principals {
			identifiers = ["ecs-tasks.amazonaws.com"]
			type = "Service"
		}
	}
}
resource "aws_iam_role" "this" { assume_role_policy = data.aws_iam_policy_document.this.json }
resource "aws_iam_role_policy_attachment" "default" {
	policy_arn  = "arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"
	role = resource.aws_iam_role.this.name
}

# * --- Step 15 ---
# * Now, we can write our ECS Task Definition which tells our cluster how to run our application.
# *
resource "aws_ecs_task_definition" "this" {
	# * A container definition may have multiple services, but ours is just our single app.
	container_definitions = jsonencode([{
		environment: [
			{ name = "MY_INPUT_ENV_VAR", value = "terraform-modified-env-var" }
		],
		essential = true,
		image = resource.docker_registry_image.this.name,
		name = local.container_name,
		portMappings = [{ containerPort = local.container_port }],
	}])
	cpu = 256
	execution_role_arn = resource.aws_iam_role.this.arn
	family = "family-of-${local.example}-tasks"
	memory = 512
	network_mode = "awsvpc"
	requires_compatibilities = ["FARGATE"]
}
```

You can view the complete workshop code (which is based off this repo) here: https://github.com/DogParkLabs/engineering-workshops/blob/main/terraform-aws-ecs-tutorial/main-complete.tfx (ignore the `.tfx` extention, that was just so Terraform would ignore the file while we ran `main.tf` for the workshop).

Nonetheless, hopefully this helps and I will update this repo code later this week.

**Comment by arminus** [comment]
Great, injecting step 14 and changing the execution_role_arn in the task definition accordingly solves the issue. Thanks for the quick response!

One more comment: Your sample app console.logs something - I guess logging needs to be defined explicitly for the task because I see nothing in the example-ecs-terraform-aws-modules-service logs... Maybe you could add that when you update this repo anyway...?

**Comment by 1Mill** [comment]
I think I just had the `console.log` to verify the service was running locally in development for testing purposes. However, it would be interesting to figure out how to get those logs out of Fargate - another future iteration for sure.

**Comment by 1Mill** [comment]
Fixed with 82e10756fbd4e93b847292ecd2e13ac5299ea2e4
